File: python/streaming.py
"""Stream a chat completion and print text as it arrives."""

import os

from dotenv import load_dotenv
from openai import OpenAI


def required(name: str) -> str:
    value = os.getenv(name, "").strip()
    if not value or value.startswith("replace-") or value.startswith("sk-your"):
        raise SystemExit(f"Set {name} in .env before running this example.")
    return value


load_dotenv()
client = OpenAI(
    api_key=required("MODB_API_KEY"),
    base_url=os.getenv("MODB_BASE_URL", "https://modbapi.com/v1").rstrip("/"),
    timeout=float(os.getenv("MODB_TIMEOUT_SECONDS", "30")),
)

stream = client.chat.completions.create(
    model=required("MODB_MODEL"),
    messages=[{"role": "user", "content": "Count from one to three."}],
    max_tokens=64,
    stream=True,
)

for chunk in stream:
    text = chunk.choices[0].delta.content if chunk.choices else None
    if text:
        print(text, end="", flush=True)
print()

File: nodejs/streaming.mjs
const apiKey = process.env.MODB_API_KEY?.trim()
const baseUrl = (process.env.MODB_BASE_URL || 'https://modbapi.com/v1').replace(/\/$/, '')
const model = process.env.MODB_MODEL?.trim()
if (!apiKey || apiKey.startsWith('sk-your')) throw new Error('Set MODB_API_KEY in ../.env.')
if (!model || model.startsWith('replace-')) throw new Error('Set MODB_MODEL in ../.env.')

const response = await fetch(`${baseUrl}/chat/completions`, {
  method: 'POST',
  headers: { Authorization: `Bearer ${apiKey}`, 'Content-Type': 'application/json' },
  body: JSON.stringify({
    model,
    messages: [{ role: 'user', content: 'Count from one to three.' }],
    stream: true,
    max_tokens: 64,
  }),
})
if (!response.ok || !response.body) throw new Error(`${response.status}: ${await response.text()}`)

const reader = response.body.pipeThrough(new TextDecoderStream()).getReader()
let buffer = ''
while (true) {
  const { value, done } = await reader.read()
  if (done) break
  buffer += value
  const lines = buffer.split('\n')
  buffer = lines.pop() || ''
  for (const line of lines) {
    if (!line.startsWith('data: ') || line === 'data: [DONE]') continue
    const chunk = JSON.parse(line.slice(6))
    const text = chunk.choices?.[0]?.delta?.content
    if (text) process.stdout.write(text)
  }
}
process.stdout.write('\n')
